.pre-commit-config.yaml

```diff
@@ -0,0 +1,37 @@
+# See https://pre-commit.com for more information
+# See https://pre-commit.com/hooks.html for more hooks
+repos:
+- repo: https://github.com/pre-commit/pre-commit-hooks
+  rev: v3.2.0
+  hooks:
+    - id: trailing-whitespace
+    - id: end-of-file-fixer
+    - id: check-yaml
+    - id: check-json
+    - id: check-toml
+    - id: pretty-format-json
+    - id: check-added-large-files
+    - id: check-case-conflict
+    - id: check-executables-have-shebangs
+    - id: check-merge-conflict
+    - id: detect-private-key
+- repo: https://github.com/hackaugusto/pre-commit-cargo
+  rev: v1.0.0
+  hooks:
+    # Allows cargo fmt to modify the source code prior to the commit
+    - id: cargo
+      name: Cargo fmt
+      args: ["+nightly", "fmt", "--all"]
+      stages: [commit]
```

Comment:
the other items will perform checks, and this one will mutate the code. shouldn't we avoid automatic mutations (as it might create a magic box effect)?

Reply:
I'm on the opinion that checks that are deterministic and can be applied automatically should always be so. Specially something like this, the developer does not have an option of not complying with the formatting, if there is an error, it will be rejected in the CI, and there is a tool that can automatically and deterministically fix the issue, so IMO it should be applied.

The reason is that I want to fix as many issues on my local machine as possible, prior to reaching CI. This is because CI is a limited resource, and as team grows we start hitting queuing in the CI, having to wait for a job to run to find out that a static check that could be automatically fixed failed isn't great DX.

With that said, I will use on my machine the automatic formatting, but we don't have to have it in the pre-commit file if other people don't want it.